ambry-replication/src/main/java/com/github/ambry/replication/BackupCheckerThread.java

```diff
@@ -0,0 +1,222 @@
+    // If we are here, then the blob exists locally and is not deleted.
+    EnumSet<MessageInfoType> acceptableLocalBlobStates = EnumSet.of(MessageInfoType.DELETE);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.noneOf(StoreErrorCodes.class);
+    // Check local store once before logging an error
+    checkLocalStore(remoteBlob, remoteReplicaInfo, acceptableLocalBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * This method checks if permanent blob from remote on-prem server is permanent in local-store
+   * @param remoteBlob the {@link MessageInfo} that will be transformed into a TTL update
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyTtlUpdate(MessageInfo remoteBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    // If we are here, then the blob exists locally and is not ttl-updated.
+    EnumSet<MessageInfoType> acceptableLocalBlobStates = EnumSet.of(MessageInfoType.TTL_UPDATE);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.noneOf(StoreErrorCodes.class);
+    // Check local store once before logging an error
+    checkLocalStore(remoteBlob, remoteReplicaInfo, acceptableLocalBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * This method checks if an un-deleted blob from remote on-prem server has been un-deleted from local-store
+   * @param remoteBlob the {@link MessageInfo} that will be transformed into an un-delete
+   * @param remoteReplicaInfo The remote replica that is being replicated from
+   */
+  @Override
+  protected void applyUndelete(MessageInfo remoteBlob, RemoteReplicaInfo remoteReplicaInfo) {
+    // If we are here, then the blob exists locally and is not un-deleted.
+    EnumSet<MessageInfoType> acceptableLocalBlobStates = EnumSet.of(MessageInfoType.UNDELETE);
+    EnumSet<StoreErrorCodes> acceptableStoreErrorCodes = EnumSet.noneOf(StoreErrorCodes.class);
+    // Check local store once before logging an error
+    checkLocalStore(remoteBlob, remoteReplicaInfo, acceptableLocalBlobStates, acceptableStoreErrorCodes);
+  }
+
+  /**
+   * Checks if a blob from remote replica is present locally
+   * We should not be here since we are overriding fixMissingStoreKeys but still need to override in case someone
+   * makes a change in the future and we end up here.
+   * @param validMessageDetectionInputStream Stream of valid blob IDs
+   * @param remoteReplicaInfo Info about remote replica from which we are replicating
+   * @throws StoreException Throws StoreException if we reach here during local-store check
+   */
+  @Override
+  protected void applyPut(MessageSievingInputStream validMessageDetectionInputStream, RemoteReplicaInfo remoteReplicaInfo)
+      throws StoreException {
+    throw new StoreException("We should not be in applyPut since we override fixMissingStoreKeys",
+        StoreErrorCodes.Put_Not_Allowed);
```

Comment:
So how do we catch those logic problems other than throwing an error ? I introduced this error code because there was no other code that fit this case.

Reply:
How do we catch such cases, other than throwing an exception ? 
Also I can throw the exception with some other error code but nothing fits this case.

Reply:
Don't add StoreErrorCodes but throw some other exceptions like IllegalStateException as our other code.

Reply:
The function signature throws either StoreException or IOException

Reply:
ok done